### PR TITLE
MTDSA-14539: Updated validation for OffPayrollWorker

### PR DESCRIPTION
**Pull request by sean-flynn-HMRC**
in [v9.8.1. API#1935(1643) - Amend Employment Financial Details - TY (23-24)](https://confluence.tools.tax.service.gov.uk/pages/viewpage.action?pageId=561971214)

### CODE REVIEW

File: app/v1/requestParsers/validators/AmendFinancialDetailsValidator.scala [hunk 1]
```diff
@@ -77,7 +78,21 @@ class AmendFinancialDetailsValidator @Inject() (implicit currentDateTime: Curren
...
 
-    jsonFormatError ++ emptyObjectValidation
+    val offPayrollWorkerAllowedValidation = {
+      val offPayrollWorker: Option[Boolean] = requestBodyObj.flatMap(entity => entity.employment.offPayrollWorker)
+      val myTaxYear                         = TaxYear.fromMtd(data.taxYear).year
+      List(
+        offPayrollWorker match {
+          case Some(false) if (myTaxYear >= 2023) => List(RuleIncorrectOrEmptyBodyError.copy(paths = Some(Seq("/employment/offPayrollWorker"))))
+          case Some(true) if (myTaxYear >= 2023)  => NoValidationErrors
+          case _ if (myTaxYear >= 2023)           => List(RuleMissingOffPayrollWorker)
```
Comment (narendravijayarao) [inline app/v1/requestParsers/validators/AmendFinancialDetailsValidator.scala:88]: Instead of _, can we use None?

Comment (narendravijayarao) [reply app/v1/requestParsers/validators/AmendFinancialDetailsValidator.scala:88]: Also, why don't we do a multi-conditional check?? Bcoz we only need **offPayrollWorker** after 2023. So better first apply the condition on the tax year, then offPayrollWorker!!

Comment (narendravijayarao) [reply app/v1/requestParsers/validators/AmendFinancialDetailsValidator.scala:88]: how about this?

```suggestion
          List(
        (myTaxYear >= 2023, offPayrollWorker) match {
            case (true, Some(true)) => NoValidationErrors
            case (true, Some(false)) => List(RuleIncorrectOrEmptyBodyError.copy(paths = Some(Seq("/employment/offPayrollWorker"))))
            case (true, None) => List(RuleMissingOffPayrollWorker)
            case (false, Some(true)) => List(RuleNotAllowedOffPayrollWorker)
            case (false, None) => NoValidationErrors
          }
      )
```

Comment (sean-flynn-HMRC) [reply app/v1/requestParsers/validators/AmendFinancialDetailsValidator.scala]: Did you write [this](https://images-na.ssl-images-amazon.com/images/I/51VhK-eG-AL._CR0,66,523,523_UX175.jpg)? 

Your suggestion is much nicer so I've updated the code accordingly. I had to make the final case _ to ensure the match statement was exhaustive though

**Review by narendravijayarao — APPROVED** [review]

**Pull request merged**